server/datastore/mysql/migrations/tables/20240703193943_AddBundleIdentifierToSoftwareTitles.go

```diff
@@ -0,0 +1,92 @@
+package tables
+
+import (
+	"database/sql"
+	"fmt"
+)
+
+func init() {
+	MigrationClient.AddMigration(Up_20240703193943, Down_20240703193943)
+}
+
+func Up_20240703193943(tx *sql.Tx) error {
+	if columnExists(tx, "software_titles", "bundle_identifier") {
+		return nil
+	}
+
+	_, err := tx.Exec(`
+		ALTER TABLE software_titles 
+		ADD COLUMN bundle_identifier varchar(255) COLLATE utf8mb4_unicode_ci DEFAULT NULL`)
+	if err != nil {
+		return fmt.Errorf("adding bundle_identifier to software_titles: %w", err)
+	}
+
+	_, err = tx.Exec(`
+	      UPDATE
+		software_titles st,
+		software s
+	      SET
+		st.bundle_identifier = NULLIF(s.bundle_identifier, '')
+	      WHERE
+		s.title_id = st.id
+	`)
+	if err != nil {
+		return fmt.Errorf("adding bundle_identifier to existing software_titles rows: %w", err)
+	}
+
+	// To account for an edge case of two software_installers pointing to
+	// different software_titles that have the same bundle_identifier now.
+	_, err = tx.Exec(`
+	    UPDATE software_installers si
+	    JOIN (
+		SELECT id, bundle_identifier
+		FROM software_titles
+	    ) st ON si.title_id = st.id
+	    JOIN (
+		SELECT MIN(id) as min_id, bundle_identifier
+		FROM software_titles
+		GROUP BY bundle_identifier
```

Comment:
I'm not sure off the top of my head, will that group all NULL bundle-identifiers together? Or that doesn't matter because it is joined on bundle_identifier so the NULLs are ignored anyway?

Reply:
> Or that doesn't matter because it is joined on bundle_identifier so the NULLs are ignored anyway?

that was my thought, I will add a test case just so we're sure